### PR TITLE
Lazy loading: feature toggle

### PR DESCRIPTION
**Pull request by bwindels**
Depends on https://github.com/matrix-org/matrix-js-sdk/pull/687 for detecting server support.

can't logout as we planned as that would have cleared the localstorage, clearing the setting that was just disabled. Putting it in account data also seemed hard, as that's not supported for features.
So the solution I took was to not logout, just stop the client, clear indexeddb and restart the app with the same credentials. A bit dangerous but seems to work well.

`feature_lazyloading` needs to be set to labs for the check to appear. Looks like this:

Missing: 
- [x] translations
- [x] check for server support
- [x] more testing, fix build

### CODE REVIEW

File: src/settings/controllers/LazyLoadingController.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import dis from "../../dispatcher";
+import { _t } from "../../languageHandler";
+
+export default class LazyLoadingController extends SettingController {
+    onChange(level, roomId, newValue) {
+        dis.dispatch({action: 'flush_storage_reload'});
+    }
+
+    canChangeTo(level, roomId, newValue) {
+        return new Promise((resolve) => this._showReloadDialog(resolve));
```
Comment (turt2live) [inline src/settings/controllers/LazyLoadingController.js]: This kinda feels like an abuse of setting controllers: the warning should be hooked up elsewhere in my opinion. The setting controllers are intended to be used for things like desktop notifications where the platform may not support it at all, so the feature should not be enabled.

There's some information about their intent here: https://github.com/matrix-org/matrix-react-sdk/blob/master/docs/settings.md#setting-controllers

File: src/settings/SettingsStore.js [hunk 2]
```diff
@@ -260,7 +260,7 @@ export default class SettingsStore {
      * @param {*} value The new value of the setting, may be null.
      * @return {Promise} Resolves when the setting has been changed.
      */
-    static setValue(settingName, roomId, level, value) {
+    static async setValue(settingName, roomId, level, value) {
```
Comment (turt2live) [inline src/settings/SettingsStore.js]: I'm very worried about making this async as I'm fairly confident that there are places in the project that expect this to be sync.

Comment (bwindels) [reply src/settings/SettingsStore.js]: But it already returned a promise before? The external API hasn't changed AFAIK.

File: src/settings/controllers/LazyLoadingController.js (hunk 1 above)
Comment (bwindels) [reply src/settings/controllers/LazyLoadingController.js]: Ok, I'll do it in the onChange handler for labs checks in UserSettings. As this won't stay a feature for too long hopefully, it's good not to make changes too deep :+1:

File: src/settings/SettingsStore.js (hunk 2 above)
Comment (bwindels) [reply src/settings/SettingsStore.js]: E.g. this is just about readability AFAIAC. The behaviour shouldn't have changed at all now canChangeTo is removed. I can still take it out if you prefer.

**Comment by bwindels** [comment]
Build is broken because the tests are run against `js-sdk/develop`, and this needs `js-sdk/bwindels/feature_lazyloading`. Only error is `TypeError: room.getMyMembership is not a function`

File: src/settings/SettingsStore.js (hunk 2 above)
Comment (turt2live) [reply src/settings/SettingsStore.js]: Ah, I see. It should be the same, yes. Mixing up my mental iterations of this section

**Comment by bwindels** [comment]
@dbkr I thought that would also get rid of the crypto database, but it looks like that isn't the case, so yes, that would be way better!

File: src/settings/controllers/LazyLoadingController.js [hunk 3]
```diff
@@ -15,10 +15,15 @@ limitations under the License.
...
 
 import SettingController from "./SettingController";
-import dis from "../../dispatcher";
+import MatrixClientPeg from "../../MatrixClientPeg";
+import PlatformPeg from "../../PlatformPeg";
 
 export default class LazyLoadingController extends SettingController {
-    onChange(level, roomId, newValue) {
-        dis.dispatch({action: 'flush_storage_reload'});
+    async onChange(level, roomId, newValue) {
```
Comment (bwindels) [inline src/settings/controllers/LazyLoadingController.js]: copied the code from `UserSettings` here as I the method there is used in a different use case, to not have this break when the method changes.

**Comment by bwindels** [comment]
@dbkr implemented your suggestion

**Pull request merged**